include/cifer/abe/dippe.h

```diff
@@ -0,0 +1,220 @@
+#include <amcl/fp12_BN254.h>
+#include <amcl/ecp_BN254.h>
+#include <amcl/ecp2_BN254.h>
+
+#include "cifer/data/vec.h"
+#include "cifer/internal/errors.h"
+
+/**
+* \file
+* \ingroup abe
+* \brief This is a decentralized inner-product predicate encryption
+* scheme by Yan Michalevsky and Marc Joye: "Decentralized Policy-Hiding
+* Attribute-Based Encryption with Receiver Privacy"
+*/
+
+
+/**
+ * cfe_dippe_cipher represents the ciphertext of the DIPPE scheme
+ */
+typedef struct cfe_dippe_cipher {
+    FP12_BN254   C_prime;
+    ECP_BN254*   C0;
+    ECP_BN254*   Ci;
+} cfe_dippe_cipher;
+
+/**
+ * Clears the ciphertext and frees allocated memory
+ *
+ * @param cipher A pointer to a cfe_dippe_cipher struct
+ */
+void cfe_dippe_cipher_free(cfe_dippe_cipher *cipher);
+
+/**
+ * cfe_dippe_pub_key represents the public key of the DIPPE scheme
+ */
+typedef struct cfe_dippe_pub_key {
+    ECP2_BN254   g2_sigma;
+	ECP_BN254*   g1_W_A;
+	FP12_BN254*  gt_alpha_A;
+} cfe_dippe_pub_key;
+
+/**
+ * Clears the public key and frees allocated memory
+ *
+ * @param pk A pointer to a cfe_dippe_pub_key struct
+ */
+void cfe_dippe_pub_key_free(cfe_dippe_pub_key *pk);
+
+/**
+ * cfe_dippe_sec_key represents the secret key of the DIPPE scheme
+ */
+typedef struct cfe_dippe_sec_key {
+	mpz_t        sigma;
+	cfe_vec      alpha;
+	cfe_vec      W;
+} cfe_dippe_sec_key;
+
+/**
+ * Clears the secret key and frees allocated memory
+ *
+ * @param sk A pointer to a cfe_dippe_sec_key struct
+ */
+void cfe_dippe_sec_key_free(cfe_dippe_sec_key *sk);
+
+/**
+ * cfe_dippe_sec_key represents the user secret key of the DIPPE scheme
+ */
+typedef struct cfe_dippe_user_secret_key {
+    ECP2_BN254* Ki;
+} cfe_dippe_user_secret_key;
+
+/**
+ * Clears the user secret key and frees allocated memory
+ *
+ * @param usk A pointer to a cfe_dippe_user_secret_key struct
+ */
+void cfe_dippe_user_secret_key_free(cfe_dippe_user_secret_key *usk);
+
+/**
+ * cfe_dippe represents the DIPPE scheme
+ */
+typedef struct cfe_dippe {
+    unsigned int assump_size;
+    mpz_t        p;
+    ECP_BN254*   g1_A;
+    ECP_BN254*   g1_UA;
+} cfe_dippe;
+
+/**
+ * Clears the scheme and frees allocated memory
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ */
+void cfe_dippe_free(cfe_dippe *dippe);
+
+/**
+ * Initializes the DIPPE scheme and generates the global parameters
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param assump_size The size of the underlying assumption
+ */
+void cfe_dippe_setup(cfe_dippe *dippe, unsigned int assump_size);
+
+/**
+ * Initializes an authority and generates its public and private key
+ *
+ * @param dippe A pointer to a cfe_dippe struct
+ * @param pk A pointer to a cfe_dippe_pub_key struct; Represents the public key that is about to be populated
+ * @param sk A pointer to a cfe_dippe_sec_key struct; Represents the secret key that is about to be populated
+ */
+void cfe_dippe_authsetup(cfe_dippe *dippe, cfe_dippe_pub_key *pk, cfe_dippe_sec_key *sk);
```

Comment:
There are a few problems that occur in almost all of the functions here.
- Order of parameters: We follow [GMP's convention](https://gmplib.org/manual/Parameter-Conventions.html#Parameter-Conventions), which is `func_name(return_value, op1, op2...)`. In this case, it should be `cfe_dippe_auth_init(cfe_dippe_pub_key *pk, cfe_dippe_sec_key *sk, cfe_dippe *dippe)
- Parameter initialization: Again, following [another GMP's convention](https://gmplib.org/manual/Variable-Conventions.html#Variable-Conventions), we require that all function parameters are initialized. Thus, we have `_init` functions for each defined struct which need to be called by the user before calling the  Look at other schemes, e.g. abe/fame or abe/gpsw as examples. Only the `_ìnit` functions are allowed to allocate memory they don't free (for this, each `_init` function has a `_free` function, which you've already implemented).

Please apply these two points to all of your implemented functions.

Reply:
I swapped the order of function parameters, so that the return value is now always on the first position. Functions that allocates memory now also have the `_init` within their names.